crates/ruff/src/rules/flake8_todo/rules.rs

```diff
@@ -0,0 +1,525 @@
+        ),
+    })
+}
+
+/// Check that the tag is valid. This function modifies `diagnostics` in-place.
+fn check_for_tag_errors(
+    tag: &Tag,
+    diagnostics: &mut Vec<Diagnostic>,
+    autofix: flags::Autofix,
+    settings: &Settings,
+) {
+    if tag.content != "TODO" {
+        // TDO-001
+        diagnostics.push(Diagnostic::new(
+            InvalidTodoTag {
+                tag: tag.content.to_string(),
+            },
+            tag.range,
+        ));
+
+        if tag.content.to_uppercase() == "TODO" {
+            // TDO-006
+            let mut invalid_capitalization = Diagnostic::new(
+                InvalidCapitalizationInTodo {
+                    tag: tag.content.to_string(),
+                },
+                tag.range,
+            );
+
+            if autofix.into() && settings.rules.should_fix(Rule::InvalidCapitalizationInTodo) {
+                invalid_capitalization.set_fix(Fix::unspecified(Edit::replacement(
+                    "TODO".to_string(),
+                    tag.range.start(),
+                    tag.range.end(),
+                )));
+            }
+
+            diagnostics.push(invalid_capitalization);
+        }
+    }
+}
+
+/// Checks for "static" errors in the comment - missing colon, missing author, etc. This function
+/// modifies `diagnostics` in-place.
+fn check_for_static_errors(
+    comment: &str,
+    comment_range: TextRange,
+    tag: &Tag,
+    diagnostics: &mut Vec<Diagnostic>,
+) {
+    // Relative offset of the current character from the start of the comment.
+    let mut relative_offset: usize = usize::try_from(tag.range.end() - comment_range.start())
+        .ok()
+        .unwrap();
+    let mut comment_chars = comment.chars().skip(relative_offset).peekable();
+    // Absolute offset of the comment's author block from the start of the file.
+    let mut author_range: Option<TextRange> = None;
+    // Absolute offset of the comment's colon from the start of the file.
+    let mut colon_offset: Option<TextSize> = None;
+
+    // An "author block" must be contained in parentheses, like "(ruff)". To check if it exists,
+    // we can check the first non-whitespace character after the tag. If that first character is a
+    // left parenthesis, we can say that we have an author's block.
+    for char in comment_chars.by_ref() {
+        relative_offset += 1;
+        if char.is_whitespace() {
+            continue;
+        }
+
+        // We can guarantee that there's no author if the colon directly follows the TODO tag.
+        if char == ':' {
+            colon_offset =
+                Some(comment_range.start() + TextSize::try_from(relative_offset).ok().unwrap());
+            break;
+        }
+
+        if char == '(' {
+            author_range = Some(TextRange::at(
+                comment_range.start() + TextSize::try_from(relative_offset).ok().unwrap(),
+                TextSize::new(1),
+            ));
+        }
+
+        break;
+    }
```

Comment:
Nit: I think you can simplify this somewhat by using `trim_start` and `starts_with`

```rust
let trimmed = comment_rest.trim_start();
let content_offset = comment_rest.text_len() = trimmed.text_len(); // whitespace length

let author_end = content_offset + if trimmed.starts_with('(') {
	if let Some(end_index) = trimmed.find(')') {
		// Test if it is not empty? 
		TextSize::try_from(end_index + 1).unwrap()
	} else {
		// Missing closing ')'
		// Create a diagnostic?
		trimmed.text_len()
	}
} else {
	diagnostics.push(Diagnostic::new(
          MissingAuthorInTodo,
          TextRange::at(content_offset, TextSize::new(1)),
      ));

	TextSize::new(0)
};

let after_author = &comment_rest[usize::from(author_end)..];

let after_colon = if let Some((_colon, after_colon)) = after_author.split_once(':') {
	if after_colon.starts_with(' ') {
		// all good
		&after_colon[1..]
	} else {
		diagnostics.push(Diagnostic::new(
              MissingSpaceAfterColonInTodo,
              TextRange::at(range, TextSize::new(1)),
          ));
    }
} else {
	// No colon
	diagnostics.push(Diagnostic::new(
          MissingColonInTodo,
          TextRange::at(adjusted_colon_position, TextSize::new(1)),
      ));
};

if after_colon.is_empty() {
	diagnostics.push(Diagnostic::new(
          MissingTextInTodo,
          TextRange::at(comment_range.end(), TextSize::new(1)),
      )),
}
```

I haven't updated all ranges.

Reply:
Really appreciate this detailed suggestion. Makes it much easier to reason about than the `char` iterator approach I was using